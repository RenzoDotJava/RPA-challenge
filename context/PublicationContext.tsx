import { createContext } from "react";
import { IUserPublications } from "../interfaces/publication.interface";

type PublicationContextProps = {
	filteredUsersPublications: IUserPublications[];
	search: string;
	loading: boolean;
	getUsersPublications: (refresh: boolean) => Promise<IUserPublications[]>;
	searchFilter: (text: string) => void;
	onRefreshPublications: () => void;
};

export const PublicationContext = createContext<PublicationContextProps>(
	{} as PublicationContextProps
);
