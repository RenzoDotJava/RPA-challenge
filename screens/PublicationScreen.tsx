import {
	StyleSheet,
	SafeAreaView,
	Platform,
	StatusBar,
} from "react-native";
import Header from "../components/Header";
import PublicationList from "../components/PublicationList";

const PublicationScreen: React.FC = () => {
	return (
		<SafeAreaView style={styles.androidSafeArea}>
			<Header />
			<PublicationList />
		</SafeAreaView>
	);
};

const styles = StyleSheet.create({
	androidSafeArea: {
		flex: 1,
		backgroundColor: "black",
		paddingTop: Platform.OS === "android" ? StatusBar.currentHeight : 0,
	},
});

export default PublicationScreen;
