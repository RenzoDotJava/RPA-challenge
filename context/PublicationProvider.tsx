import AsyncStorage from "@react-native-async-storage/async-storage";
import { useState, useEffect, useCallback } from "react";
import { IPublication, IUserPublications } from "../interfaces/publication.interface";
import { PublicationContext } from "./PublicationContext";
import { PublicationService } from "../services/publicationService";

export const PublicationProvider: React.FC<any> = ({ children }) => {
	const [usersPublications, setUsersPublications] = useState<IUserPublications[]>([]);
	const [filteredUsersPublications, setFilteredUsersPublications] = useState<IUserPublications[]>([]);
  const [search, setSearch] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);

  const publicationService: PublicationService = new PublicationService();

  useEffect(() => {
		let isCanceled: boolean = false;

		getUsersPublications().then((res: IUserPublications[]) => {
			if (isCanceled) return;

			setUsersPublications(res);
			setFilteredUsersPublications(res);
		});

		return () => {
			isCanceled = true;
		};
	}, []);

  const groupPublicationsByUserId = (
    publications: IPublication[]
  ): IUserPublications[] => {
    return publications.reduce((acc: IUserPublications[], val: IPublication) => {
      let publication: IPublication = {
        id: val.id,
        title: val.title,
        body: val.body.replace(/\n|\r/g, ""),
      };
  
      const existingUser: IUserPublications | undefined = acc.find(
        (group: IUserPublications) => group.userId === val.userId
      );
  
      if (existingUser) {
        existingUser.data.push(publication);
      } else {
        acc.push({
          userId: val.userId,
          data: [publication],
        });
      }
  
      return acc;
    }, []);
  };

  const getPublicationsByUsersAmount = (
    amount: number = 5
  ): Promise<IUserPublications[]> => {
    let indexes: number[] = Array.from({ length: amount }, (val, i) => i);
    return Promise.all(
      indexes.map((index) => {
        return publicationService.getPublicationsByUserId(index + 1);
      })
    ).then(async (res: IPublication[][]) => {
      const groupedPublications: IUserPublications[] =
        groupPublicationsByUserId(res.flat());
  
      await AsyncStorage.setItem(
        "usersPublications",
        JSON.stringify(groupedPublications)
      );
  
      return groupedPublications;
    });
  };

  const getUsersPublications = async (
    refresh: boolean = false
  ): Promise<IUserPublications[]> => {
    let json: string | null = "";
    let usersPublications: IUserPublications[] = [];
  
    if (!refresh) json = await AsyncStorage.getItem("usersPublications");
  
    if (!json || json == "") {
      await getPublicationsByUsersAmount().then((res) => {
        usersPublications = res;
      });
    } else {
      usersPublications = JSON.parse(json);
    }
  
    return usersPublications;
  };

  const searchFilter = (text: string) => {
		if (text.trim() == "") {
			setFilteredUsersPublications(usersPublications);
		} else {
			const newPublications: IUserPublications[] = usersPublications
				.filter((item) =>
					item.data.some((d) => d.title.includes(text.toLowerCase()))
				)
				.map((item) => ({
					userId: item.userId,
					data: item.data.filter((d) =>
						d.title.includes(text.toLowerCase())
					),
				}));

			setFilteredUsersPublications(newPublications);
		}

		setSearch(text);
	};

  const onRefreshPublications = useCallback(() => {
		setLoading(true);
		getUsersPublications(true)
			.then((res: IUserPublications[]) => {
				setUsersPublications(res);
				setFilteredUsersPublications(res);
			})
			.catch(() => {
				setUsersPublications([]);
				setFilteredUsersPublications([]);
			})
			.finally(() => {
				setSearch("");
				setLoading(false);
			});
	}, []);

	return (
		<PublicationContext.Provider
			value={{
				filteredUsersPublications,
        search,
        loading,
        getUsersPublications,
        searchFilter,
        onRefreshPublications
			}}
		>
			{children}
		</PublicationContext.Provider>
	);
};
