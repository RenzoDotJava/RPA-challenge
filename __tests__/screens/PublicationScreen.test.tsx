import { render, act, waitFor, screen, fireEvent } from "@testing-library/react-native";
import PublicationScreen from "../../screens/PublicationScreen";
import { PublicationContext } from "../../context/PublicationContext";

describe("Testing <PublicationScreen />", () => {
	beforeEach(() => jest.clearAllMocks());

	it("Matching snapshot", () => {
		render(
			<PublicationContext.Provider
				value={{
					filteredUsersPublications: [],
					search: "",
					loading: false,
					getUsersPublications: jest.fn(),
					searchFilter: jest.fn(),
					onRefreshPublications: jest.fn()
				}}
			>
				<PublicationScreen />
			</PublicationContext.Provider>
		);

		act(() => {
			expect(screen).toMatchSnapshot();
		});

		waitFor(() => {
			expect(screen.getByTestId("publication-list")).toBeDefined();
		});
	});
});
