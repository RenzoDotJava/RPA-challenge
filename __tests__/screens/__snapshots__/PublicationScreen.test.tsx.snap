// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <PublicationScreen /> Matching snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "black",
      "flex": 1,
      "paddingTop": 0,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "black",
        "flexDirection": "row",
        "paddingBottom": 8,
      }
    }
    testID="header"
  >
    <View
      style={
        {
          "borderBottomColor": "white",
          "borderBottomWidth": 2,
          "flexDirection": "row",
          "marginBottom": 1,
          "marginHorizontal": 12,
          "marginTop": 10,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        style={
          {
            "color": "white",
            "flex": 1,
            "paddingBottom": 15,
          }
        }
        testID="search-bar"
        value=""
      />
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#CCCCCC",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <EmptyList
          text="No hay publicaciones"
        />
      }
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          colors={
            [
              "#32373A",
            ]
          }
          onRefresh={[MockFunction]}
          refreshing={false}
          tintColor="#32373A"
        />
      }
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      testID="publication-list"
    >
      <RCTRefreshControl />
      <View>
        <View
          style={
            {
              "flex": 1,
              "marginTop": 15,
            }
          }
        >
          <Text
            style={
              {
                "fontFamily": "Arial",
                "fontSize": 20,
                "fontWeight": "500",
                "textAlign": "center",
              }
            }
          >
            No hay publicaciones
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
